README.md

```diff
@@ -10,26 +10,222 @@ Distributed Consensus Algorithms often take the form of a replicated state machi
 
 Two well known Distributed Consensus Algorithms are Paxos and Raft. Paxos is used in systems like [Chubby](http://research.google.com/archive/chubby.html) by Google, and Raft is used in things like [`tikv`](https://github.com/pingcap/tikv) or [`etcd`](https://github.com/coreos/etcd/tree/master/raft). Raft is generally seen as a more understandable and simpler to implement than Paxos.
 
-## Links for Further Research
+## Design
 
-* [The Raft site](https://raftconsensus.github.io/)
-* [The Secret Lives of Data - Raft](http://thesecretlivesofdata.com/raft/)
-* [Raft Paper](http://ramcloud.stanford.edu/raft.pdf)
-* [Raft Dissertation](https://github.com/ongardie/dissertation#readme)
-* [Raft Refloated](https://www.cl.cam.ac.uk/~ms705/pub/papers/2015-osr-raft.pdf)
+Raft replicates the state machine through logs. If you can ensure that all the machines have the same sequence of logs, after applying all logs in order, the state machine will reach a consistent state. 
 
-## Using It
+A complete Raft model contains 4 essential parts: 
 
-To include this project as a dependency:
+1. Consensus Module, the core consensus algorithm module;
 
-```
+2. Log, the place to keep the Raft logs;
+
+3. State Machine, the place to save the user data;
+
+4. Transport, the network layer for communication.
+
+![The design of raft-rs](media/the-design-of-raft-rs.png) 
+
+> Note: This Raft implementation in Rust includes the core Consensus Module only, not the other parts. The core Consensus Module in raft-rs is customizable, flexible, and resilient. You can directly use raft-rs, but you will need to build your own Log, State Machine and Transport components. 
+
+## How to use raft-rs
+
+Before using Raft-rs, include this project as a dependency:
+
+```bash
 [dependencies]
 raft = "0.2"
 ```
 
-This crate is currently being used in production with [TiKV](https://github.com/pingcap/tikv), and is an actively developed project. There are still [several jobs](https://github.com/pingcap/raft-rs/issues/35), including writing extensive documentation and creating more examples, that need to be finished before reaching 1.0.
+### Step 1: Create the Raft node
+
+You can use `RawNode::new()` to create the Raft node. To create the Raft node, you need to build a Log component, which is called Raft Storage in raft-rs, and then configure the Raft node.
+
+1. Build Raft Storage
+
+	Raft Storage saves all the information about the current Raft implementation, including Raft Log, commit index, the leader to vote for, etc.
+	
+	Storage is a trait defined in [storage.rs](src/storage.rs). See the following trait interfaces in detail:
+	
+	* `initial_state` is called when Raft is initialized. This interface will return a `RaftState` which contains `HardState` and `ConfState`;
```

Comment:
These definitions should be present in [the docs](https://docs.rs/raft/0.2.0/raft/storage/trait.Storage.html). I would reccomend moving these there (some are already present) and removing them from here to reduce future maintenance burden.